### ISSUE TITLE
Provide a migration command

### ISSUE THREAD

**Issue by erkinalp** [Half-Shot/matrix-appservice-discord#681]
Allow community administrators to migrate a community from Discord to matrix wholesale.

Inputs: Discord guild ID, matrix space address, identity provider (a matrix identity server)
Outputs: freshly created matrix rooms with all users invited according to permissions
Steps of operation (the ordering is significant):
* Create empty matrix rooms corresponding to channels with appropriate permissions
* Backfill messages to respective matrix rooms
* Invite matrix users with corresponding identities to those fresh matrix rooms
* Add those fresh chat rooms to the supplied space room
* DM the users to inform on migration
* Deop all the Discord users apart from the owner (order: remove roles from all users apart from the bridge and the owner, delete overwrites)
* Prune users offline for 1 day, including those with roles
* Delete all Discord roles
* Delete all messages from Discord side, from oldest to newest (this step is not redundant as channel delete does not immediately delete all the messages)
* Delete all Discord channels
* Kick all Discord users apart from the owner

Option switches:
* No-delete: do not delete the Discord side.
* Delay-delete: delete the Discord side past a specified time after pruning the users.
* Use existing matrix chat rooms
* Create a fresh space room

**Comment by HansJK** [comment]
We really need this!

Option switches +:
* Choose not to prune users offline for 1 day, including those with roles.
* Choose not to delete roles.

**Comment by erkinalp** [comment]
> Choose not to prune users offline for 1 day, including those with roles.
> Choose not to delete roles.

No-delete would do the combination of both, in addition to not deleting the messages and channels.

**Comment by erlend-sh** [comment]
Some prior art:
https://github.com/Linen-dev/linen.dev
https://github.com/Linen-dev/linen.dev/tree/main/apps/web/services/discord
